pmd-java/etc/grammar/Java.jjt

```diff
@@ -1872,11 +1876,13 @@ void Initializer() :
  * Type, name and expression syntax follows.
  */
 void Type():
-{}
 {
-   LOOKAHEAD(2) ReferenceType()
- |
-   PrimitiveType()
+    Token t;
+}
+{
+  LOOKAHEAD( { jdkVersion >= 10 && isKeyword("var") } ) t=<IDENTIFIER> {jjtThis.setImage(t.image); jjtThis.setTypeInferred(true); }
```

Comment:
This is actually more problematic than I thought. 
* This allows `var` in many more contexts than the JLS -> not problematic per se, but could be better
* When `var` is used as a type for a local variable declaration, I think the most consistent behaviour would be to have *no Type node* in the LocalVariableDeclaration, and rather set `isTypeInferred` on the VariableDeclaratorId. This would be consistent with eg implicitly typed lambda parameters: 

https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTVariableDeclaratorId.java#L133-L146

  Having a `isTypeInferred` on the Type node is a bit weird, and it makes the grammar of this node more fragile, since we can't expect a Type node to have at least one children anymore...

I'd suggest that the `var` grammar be moved to the LocalVariableDeclaration production:
https://github.com/pmd/pmd/blob/master/pmd-java/etc/grammar/Java.jjt#L2334

You could create a `#void` production matching [this JLS segment](https://docs.oracle.com/javase/specs/jls/se10/html/jls-14.html#jls-LocalVariableType), factorising the lookup for either Type or `var`, and prepare the day we implement support for [JEP 323](http://openjdk.java.net/jeps/323) (Java 11).

Reply:
> Imho, adding this additional level (in any way) would make the AST much more incompatible.

Ok I see your point. It's true that not having the guarantee of having a Type as a child of LocalVarDecl could break some rules. 

I find it's a shame though, since our Type production was exactly the UnannType of the JLS until now. I think, we'll have to roll back these changes in 7.0.0, and introduce LocalVariableType then...

I still think this change introduces a complication for lambda parameters. Using a node for `var` in that context forces us to treat differently the case where the parameters are just identifiers and the case where they use `var`. This seems to me slightly inconsistent and awkward to support, since the two represent the same thing.

One last suggestion then: I think, what would help make the transition come 7.0.0, is to pretend like this Type node doesn't exist in case the type is inferred in the API of e.g. ASTResource, ASTVariableDeclaratorId. The node would still be reachable by rules which use raw traversal methods, such that they wouldn't break, but methods from the API (`getTypeNode`) would behave exactly like they would when we get rid of the node. This would make the API consistent at least on a high level, and would prepare the future. I wrote a [bit of doc](https://github.com/pmd/pmd/pull/1073/files#diff-7233d050a898d65d5c8d4b2383ce9e3eR140) that goes in that sense back then. If we don't do that then we need to update it

Reply:
I wouldn't care about breaking rules by not guaranteeing Type node to have at least one children anymore. People using Java 9 would see everything working just as fine. Custom rules would have to be accommodated for Java 10 to make them ready. At most we could keep Java 9 as the default version and have users switch to Java 10 to avoid unpleasant surprises.

I fear we are accumulating too much tech debt for PMD 7.0.0, and making the cost of upgrade for third parties a lot higher by introducing these compromises... just my 2 cents 'though